lib/web/fetch/index.js

```diff
@@ -126,8 +127,53 @@ class Fetch extends EE {
   }
 }
 
+// hoisting subscribers variables
+const subscribers = {
+  start: false,
+  end: false,
+  asyncStart: false,
+  asyncEnd: false,
+  fetchError: false
+}
+
+// publishIfSubscribed will check if there are subscribers to the channel
+// add set additional context before publishing the event
+function publishIfSubscribed (name, context, error) {
+  if (subscribers[name]) {
+    if (error) context.error = error
+    channels[name].publish(context)
+    if (error) context.error = null
+  }
+}
+
+// handlePromise will handle both fulfilled/rejected promises from p
+// for diagnostics channel it will then check for subscribers and
+// publish events accordingly
+function handlePromise (p, context) {
+  return p.promise.then(
+    (success) => {
+      publishIfSubscribed('asyncStart', context)
+      publishIfSubscribed('asyncEnd', context)
+      return success
+    },
+    (failure) => {
+      publishIfSubscribed('asyncStart', context, failure)
+      publishIfSubscribed('asyncEnd', context, failure)
+      return Promise.reject(failure)
+    })
+}
```

Comment:
I still dont understand what asyncStart and asyncEnd are used for?!
Also why are they emitted the same time?

Reply:
In the case of callbacks the `asyncStart` and `asyncEnd` happen before and after the callback. In the case of promises, there is no distinct "end" condition short of it being garbage collected so it just runs the `asyncEnd` immediately. If you were using the `tracingChannel` interface directly it would handle this edge case for you, but because you're opting to use individual channels directly that edge case needs to be covered manually.